File: CSharp/CommandMessenger/Queue/CollapseCommandStrategy.cs
namespace CommandMessenger
{
    public class CollapseCommandStrategy : CommandStrategy
    {
        public CollapseCommandStrategy(Command command) : base(command)
        {}

        public override void Enqueue()
        {
            //Console.WriteLine("Enqueue {0}", CommandQueue.Count);
            // Remove a command with the same CmdId
            var index = CommandQueue.FindIndex(strategy => strategy.Command.CmdId == Command.CmdId);
            if (index < 0)
            {
                CommandQueue.Enqueue(this);
            }
            else
            {
                CommandQueue[index] = this;
            }      
        }
    }
}

File: extras/Python/cmd_messenger/queue/collapse_command_strategy.py
"""``CollapseCommandStrategy`` — replaces any pending command with the same ``cmd_id``."""
from __future__ import annotations

from .command_strategy import CommandStrategy


class CollapseCommandStrategy(CommandStrategy):
    """Avoid duplicate commands of the same id in the queue (avoids lag)."""

    def enqueue(self) -> None:
        assert self.command_queue is not None
        for i, strategy in enumerate(self.command_queue):
            if strategy.command.cmd_id == self.command.cmd_id:
                self.command_queue[i] = self
                return
        self.command_queue.enqueue(self)
